cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/ClusterOperation.java

```diff
@@ -1,17 +1,123 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package io.strimzi.controller.cluster.operations;
 
 import io.strimzi.controller.cluster.K8SUtils;
+import io.strimzi.controller.cluster.resources.AbstractCluster;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.shareddata.Lock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
 
-public abstract class ClusterOperation implements Operation<K8SUtils> {
-    protected final String namespace;
-    protected final String name;
+/**
+ * Abstract cluster creation, for a generic cluster type {@code C}.
+ * This class applies the template method pattern, first obtaining the desired cluster configuration
+ * ({@link Op#getCluster(K8SUtils, String, String)}),
+ * then creating resources to match ({@link Op#futures(K8SUtils, String, AbstractCluster)}).
+ *
+ * This class manages a per-cluster-type and per-cluster locking strategy so only one operation per cluster
+ * can proceed at once.
+ * @param <C>
+ */
+public abstract class ClusterOperation<C extends AbstractCluster> {
 
+    private static final Logger log = LoggerFactory.getLogger(KafkaClusterOperation.class.getName());
```

Comment:
The logger should probably use `Clusteroperation` class?

Reply:
Well spotted, thanks.